### PR TITLE
GA-18: Add new access_type_roles table

### PR DESCRIPTION
**Pull request by helenbirdhmcts**
**Before creating a pull request make sure that:**

- [ x] commit messages are meaningful and follow good commit message guidelines
- [ x] README and other documentation has been updated / added (if needed)
- [ x] tests have been updated / new tests has been added (if needed)
- [x ] the `keep-helm` label has been added, if the helm release should be persisted after a successful build

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/GA-18

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/entity/AccessTypeRolesEntity.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    private LocalDateTime liveFrom;
+
+    @Column(name = "live_to")
+    private LocalDateTime liveTo;
+
+    @ManyToOne(fetch = LAZY)
+    @JoinColumn(name = "case_type_id", nullable = false)
+    private CaseTypeEntity caseTypeId;
+
+    @Column(name = "access_type_id")
```
Comment (sanjayparekh-hmcts) [inline repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/entity/AccessTypeRolesEntity.java:29]: I think we'd probably need to add nullable = false to some more of these fields - see required on JIRA

File: repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/model/AccessTypeRolesField.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    private String organisationProfileId;
+    private Boolean accessMandatory;
+    private Boolean accessDefault;
+    private Boolean display;
+    private String description;
+    private String hint;
+    private Integer displayOrder;
+    private String organisationalRoleName;
+    private String groupRoleName;
+    private String organisationPolicy_field;
```
Comment (sanjayparekh-hmcts) [inline repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/model/AccessTypeRolesField.java:29]: check the naming on this - you have applied a JsonNamingStrategy

File: repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql [hunk 3]
```diff
@@ -0,0 +1,60 @@
+--
+-- Name: access_type_roles; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE public.access_type_roles (
+                                        id integer NOT NULL,
+                                        live_from date NOT NULL,
```
Comment (sanjayparekh-hmcts) [inline repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:7]: not sure this col needs to be mandatory - JIRA needs updating?

File: repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ALTER SEQUENCE public.access_type_roles_id_seq OWNED BY public.access_type_roles.id;
+
+--
+-- Name: display_group fk_access_type_roles_id; Type: FK CONSTRAINT; Schema: public; Owner: -
+--
+
+ALTER TABLE ONLY public.access_type_roles
+    ADD CONSTRAINT fk_access_type_roles_id FOREIGN KEY (case_type_id) REFERENCES public.case_type(id);
+
+--
```
Comment (sanjayparekh-hmcts) [inline repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:49]: do we also need this kinda unique constraint too? some of the other tables have it

ALTER TABLE ONLY public.case_field
    ADD CONSTRAINT fk_case_field_case_type_id FOREIGN KEY (case_type_id) REFERENCES public.case_type(id);

**Review by sanjayparekh-hmcts — CHANGES_REQUESTED** [review]

File: repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/entity/AccessTypeRolesEntity.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+import javax.persistence.JoinColumn;
+
+import java.io.Serializable;
+import java.time.LocalDateTime;
+
+import static javax.persistence.FetchType.LAZY;
+import static javax.persistence.GenerationType.IDENTITY;
+
+@Table(name = "access_type_roles")
+public class AccessTypeRolesEntity implements Serializable {
```
Comment (abul-syed) [inline repository/src/main/java/uk/gov/hmcts/ccd/definition/store/repository/entity/AccessTypeRolesEntity.java:19]:
```suggestion
@Entity
public class AccessTypeRolesEntity implements Serializable {
```

Please also add JPA `@Entity` annotation.

File: repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql (hunk 4 above)
Comment (helenbirdhmcts) [reply repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:49]: little confused?  I think we need to FK as a reference to the case_type table

File: repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql (hunk 3 above)
Comment (helenbirdhmcts) [reply repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:7]: Spoke with Suman, live_from is marked as required here: https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1712500134#GroupAccessScope-AccessTypeRoles

Comment (helenbirdhmcts) [reply repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:7]: after further discussion agreed to make it non mandatory.

File: repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql (hunk 4 above)
Comment (sanjayparekh-hmcts) [reply repository/src/main/resources/db/migration/V20231121_18__GA-18_add_access_type_roles_table.sql:49]: sorry, got a little confused. you have the FK above now so thats gonna be enough
may want to change the constraint name from 
fk_access_type_roles_id
to
fk_access_type_roles_case_type_id

**Review by sanjayparekh-hmcts — CHANGES_REQUESTED** [review]

**Review by sanjayparekh-hmcts — APPROVED** [review]

**Review by patelila — COMMENTED** [review]
It Looks good, I did notice about the organisation_policy_field in Ticket does not mention Text limit, it may be an oversight. But then noticed the comments in ticket.

**Review by RebeccaBaker — APPROVED** [review]

**Comment by JamesCollettCGI** [comment]
PIPELINE: https://build.hmcts.net/job/HMCTS_a_to_c/job/ccd-definition-store-api/view/change-requests/job/PR-1420/

**Pull request merged**